pgml-sdks/python/pgml/pgml/collection.py

```diff
@@ -0,0 +1,771 @@
+import hashlib
+import json
+import uuid
+import os
+from .dbutils import run_drop_or_delete_statement, run_create_or_insert_statement, run_select_statement
+
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+
+FORMAT = "%(message)s"
+logging.basicConfig(
+    level=os.environ.get("LOGLEVEL", "ERROR"),
+    format=FORMAT,
+    datefmt="[%X]",
+    handlers=[RichHandler()],
+)
+log = logging.getLogger("rich")
+
+"""
+Collection class to store tables for documents, chunks, models, splitters, and embeddings
+"""
+
+class Collection:
+    def __init__(self, pool: ConnectionPool, name: str) -> None:
+        """
+        The function initializes an object with a connection pool and a name, and creates several tables
+        while registering a text splitter and a model.
+
+        :param pool: `pool` is an instance of `ConnectionPool` class which manages a pool of database
+        connections
+        :type pool: ConnectionPool
+        :param name: The `name` parameter is a string that represents the name of an object being
+        initialized. It is used as an identifier for the object within the code
+        :type name: str
+        """
+        self.pool = pool
+        self.name = name
+        self._create_documents_table()
+        self._create_splitter_table()
+        self._create_models_table()
+        self._create_chunks_table()
+        self._create_transforms_table()
+        self.register_text_splitter()
+        self.register_model()
+
+    def _create_documents_table(self) -> None:
+        """
+        This function creates a table and indexes for storing documents in a PostgreSQL database.
+        """
+        self.documents_table = self.name + ".documents"
+        conn = self.pool.getconn()
+
+        create_schema_statement = "CREATE SCHEMA IF NOT EXISTS %s" % self.name
+        run_create_or_insert_statement(conn, create_schema_statement)
+        create_table_statement = (
+            "CREATE TABLE IF NOT EXISTS %s (\
+                                id          serial8 PRIMARY KEY,\
+                                created_at  timestamptz NOT NULL DEFAULT now(),\
+                                document    uuid NOT NULL,\
+                                metadata    jsonb NOT NULL DEFAULT '{}',\
+                                text        text NOT NULL,\
+                                UNIQUE (document)\
+                                );"
+            % self.documents_table
+        )
+        run_create_or_insert_statement(conn, create_table_statement)
+
+        create_index_statement = (
+            "CREATE INDEX CONCURRENTLY IF NOT EXISTS \
+                                    created_at_index ON %s \
+                            (created_at);"
+            % (self.documents_table)
+        )
+        run_create_or_insert_statement(conn, create_index_statement, autocommit=True)
+
+        create_index_statement = (
+            "CREATE INDEX CONCURRENTLY IF NOT EXISTS \
```

Comment:
Don't need to create an index here on `document`, `UNIQUE` does it already automatically. So I think you end up with two indexes on the same field.